client/consensus/babe/src/rpc.rs

```diff
@@ -0,0 +1,235 @@
+			.pool_size(1)
+			.create()
+			.unwrap();
+		Self {
+			client,
+			shared_epoch_changes,
+			keystore,
+			babe_config,
+			threadpool,
+		}
+	}
+}
+
+impl<B, C> Babe for BabeRPC<B, C>
+	where
+		B: BlockT,
+		C: ProvideRuntimeApi<B> + SelectChain<B> + HeaderBackend<B>
+		+ HeaderMetadata<B, Error=BlockChainError> + 'static,
+		C::Api: BabeApi<B>,
+		<<C as ProvideRuntimeApi<B>>::Api as sp_api::ApiErrorExt>::Error: fmt::Debug
+{
+	fn epoch_authorship(&self) -> BoxFuture<Vec<SlotAuthorship>> {
+		let (
+			babe_config,
+			keystore,
+			shared_epoch,
+			client,
+		) = (
+			self.babe_config.clone(),
+			self.keystore.clone(),
+			self.shared_epoch_changes.clone(),
+			self.client.clone(),
+		);
+		let (tx, rx) = oneshot::channel();
+
+		let future = async move {
+			let header = client.best_chain().map_err(Error::Consensus)?;
+			let epoch_start = client.runtime_api()
+				.current_epoch_start(&BlockId::Hash(header.hash()))
+				.map_err(|err| {
+					Error::StringError(format!("{:?}", err))
+				})?;
+			let epoch = epoch_data(&shared_epoch, &client, &babe_config, epoch_start)?;
+			let (epoch_start, epoch_end) = (epoch.start_slot, epoch.end_slot());
+
+			let mut slots = vec![];
+
+			for slot_number in epoch_start..=epoch_end {
+				let epoch = epoch_data(&shared_epoch, &client, &babe_config, slot_number)?;
+				let slot = authorship::claim_slot(slot_number, &epoch, &babe_config, &keystore);
+				if let Some((claim, key)) = slot {
+					let claim = match claim {
+						BabePreDigest::Primary { vrf_output, vrf_proof, threshold, .. } => {
+							BabeClaim::Primary {
+								threshold,
+								output: Bytes(vrf_output.as_bytes().to_vec()),
+								proof: Bytes(vrf_proof.to_bytes().to_vec()),
+								key: key.public(),
+							}
+						}
+						BabePreDigest::Secondary { authority_index, .. } => {
+							BabeClaim::Secondary {
+								authority_index
+							}
+						}
+					};
+
+					slots.push(SlotAuthorship {
+						claim,
+						slot_number,
+					});
+				}
+			}
+
+			Ok(slots)
+		}.then(|result| {
+			let _ = tx.send(result).expect("receiever is never dropped; qed");
+			future::ready(())
+		}).boxed();
+
+		self.threadpool.spawn_ok(future);
+
+		Box::new(async {
+			rx.await.expect("sender is never dropped; qed")
+		}.boxed().compat())
+	}
+}
+
+/// slot authorship information
+#[derive(Debug, Deserialize, Serialize)]
+pub struct SlotAuthorship {
+	/// slot number in the epoch
+	slot_number: u64,
+	/// claim data
+	claim: BabeClaim,
+}
+
+/// Babe claim
+#[derive(Debug, Deserialize, Serialize)]
+enum BabeClaim {
+	/// a primary claim for a given slot
+	Primary {
+		/// epoch threshold
+		threshold: u128,
+		/// key that makes this claim
+		key: AuthorityId,
+		/// Vrf output
+		output: Bytes,
+		/// Vrf proof
+		proof: Bytes,
+	},
+	/// a secondary claim for a given slot
+	Secondary {
+		authority_index: AuthorityIndex,
+	},
+}
+
+/// Errors encountered by the RPC
+#[derive(Debug, err_derive::Error, derive_more::From)]
```

Comment:
Currently we don't use any crates for automatic error derivation, it seems we just use `derive_more` in some places to derive `Display` instances for errors. I don't have a strong opinion on this though.

Reply:
I think it's better to avoid it. `derive_more` should be sufficient, especially we don't really propagate that error anywhere and just always convert to `jsonrpc_core:Error`. IMHO not worth an extra dependency.